### PR TITLE
Add lookup feature flags

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds feature flags for the different configurations of the lookup argument, and tweaks the code for generating constraints to insert conditionals when in `generate_feature_flags` mode.

To support this, the `EnabledIf` expression constructor is replaced with an `IfFeature` constructor, which allows us to determine the value of both branches. This allows us to return `1` instead of `0` as the 'disabled' branch, which enables conditional multiplication. For example,
```
let table_id_combiner = JointCombiner * IfFeature(TableWidth(2), JointCombiner, 1) * ...
```
This includes some supporting refactorings to the lookup constraints, so that these kinds of expressions can be build easily.

This also adds a new feature for testing that the 'featured' layout is equivalent to the explicit version when we generate the constraints expression (c081f53779e667f62925485e534b0322da284989). This test is essential for correct interaction with the code we generate for pickles. Since CI runs the testsuite with `--all-features`, this test on all of the configurations that we have existing tests for.

-----------------

Note: In order to generate a 'stable' constraint layout regardless of the enabled features, we pad the compatibility checks in the lookup argument. This ensures that the 'runtime tables' constraint does not 'move around' (i.e. is not assigned different powers of alpha depending on the features). See commit 3243cc3e123f824bf5ded5ac10b4d79407e398d6 for the relevant change.

### CODE REVIEW

**Review by jspada — COMMENTED** [review]
This is a great enhancement.  I've left a bunch of feedback, suggestions and questions.  I'll make another pass after this.

It would be very good to get some more eyes on this because it's so important.

File: kimchi/src/circuits/expr.rs
Comment (jspada) [inline kimchi/src/circuits/expr.rs:559]: What's the second part of the returned tuple?

File: kimchi/src/circuits/expr.rs
Comment (jspada) [inline kimchi/src/circuits/expr.rs:558]: please comment what this function is doing, inputs and outputs

File: kimchi/src/circuits/lookup/lookups.rs
Comment (querolita) [reply kimchi/src/circuits/lookup/lookups.rs:131]: I guess a pattern comes with some inherent constraints, so they can be synonyms at this level

File: kimchi/src/circuits/expr.rs
Comment (querolita) [inline kimchi/src/circuits/expr.rs:1540]: Wouldn't we need to use `PolishToken::SkipIf` instead?

File: kimchi/src/circuits/expr.rs
Comment (querolita) [inline kimchi/src/circuits/expr.rs:2121]: Why is this always zero in the first branch, whereas in the other it can be either zero or `c`?

File: kimchi/src/circuits/lookup/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/lookup/constraints.rs:426]: I guess at least it needs to be 1, otherwise we're not looking anywhere, are we?

File: kimchi/src/circuits/lookup/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/lookup/constraints.rs:457]: similar as above, here it is starting from 1 instead of 0, I guess it needs to be 1 at least to make sense

File: kimchi/src/circuits/lookup/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/lookup/constraints.rs:503]: +1

File: kimchi/src/circuits/lookup/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/lookup/constraints.rs:643]: Because we only allow up to 4 lookups per row perhaps? I am guessing

**Review by querolita — APPROVED** [review]
I am giving a first pass and left some comments / questions to clarify the motivation behind some design decisions. It looks really nice to me, but perhaps we want to wait until more approvals are received / more eyes look at it before merge 😄

File: kimchi/src/circuits/expr.rs
Comment (querolita) [inline kimchi/src/circuits/expr.rs:559]: What is the boolean trying to represent? Whether it is a fully reduced expression?

File: kimchi/src/circuits/expr.rs
Comment (querolita) [inline kimchi/src/circuits/expr.rs:587]: Why can the zero expression be reduced further? Meaning, why is this a true? I guess I am not seeing what the bool here exactly means

File: kimchi/src/circuits/expr.rs
Comment (querolita) [inline kimchi/src/circuits/expr.rs:637]: Looks OK to me, looked at each branch carefully

File: kimchi/src/circuits/expr.rs
Comment (querolita) [inline kimchi/src/circuits/expr.rs:687]: Mmm interesting different behaviour for c1 and c2 here... what is the motivation for `is_enabled` to act like this?

**Review by querolita — APPROVED** [review]
I am leaving a second review with pending comments I had about the two commits referenced in the description of the PR

File: kimchi/src/linearization.rs
Comment (mimoo) [reply kimchi/src/linearization.rs:276]: fixed later

File: kimchi/src/linearization.rs
Comment (mimoo) [reply kimchi/src/linearization.rs:178]: but as you said elsewhere, a helper function for this would be nice

File: kimchi/src/linearization.rs
Comment (mimoo) [inline kimchi/src/linearization.rs:159]: this does not take into account runtime tables (or user-provided tables I think), so not sure I understand this check

File: kimchi/src/linearization.rs
Comment (mimoo) [inline kimchi/src/linearization.rs:189]: ok I'm happy I spotted that : D

File: kimchi/src/circuits/expr.rs
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:522]: ocaml-gen now supports `usize <-> int` so feel free to use `usize`

Comment (mimoo) [reply kimchi/src/circuits/expr.rs:522]: https://github.com/o1-labs/ocaml-gen/commit/586deaa0bd615e3326d4b1bb7396d57ceffbbbce

Comment (mimoo) [reply kimchi/src/circuits/expr.rs:522]: not sure I understand these feature flags at this point, is this going to panic if you try to create a pattern that does too many lookup per rows? or if you try to create a table that's too large?

File: kimchi/src/circuits/lookup/constraints.rs
Comment (mimoo) [reply kimchi/src/circuits/lookup/constraints.rs:503]: I reviewed this code once and it was the most confusing code to follow :D probably be good if we would generate parts of the spec from comments here. Definitely lots of opportunity to clean this up

Comment (mimoo) [reply kimchi/src/circuits/lookup/constraints.rs:503]: in general, https://o1-labs.github.io/proof-systems/rfcs/3-lookup.html is your friend to go through this part of the code

File: kimchi/src/linearization.rs
Comment (mimoo) [inline kimchi/src/linearization.rs:218]: this could just be `if cfg!(debug_assertions) {`, unless you think people are going to want to run this in prod

File: kimchi/src/circuits/expr.rs
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:681]: implement `Index<FeatureFlag>` on `LookupFeatures`?

File: kimchi/src/circuits/expr.rs
Comment (mimoo) [reply kimchi/src/circuits/expr.rs:687]: lookup stuff:

```rust
dummy_used = E::IfFeature(
                        FeatureFlag::LookupsPerRow((i + 1) as isize),
                        Box::new(dummy_used),
                        Box::new(E::one()),
                    );
```

**Review by mimoo — APPROVED** [review]
this is great!

File: kimchi/src/linearization.rs
Comment (mimoo) [reply kimchi/src/linearization.rs]: looks like it was fixed later

**Pull request merged**
